### PR TITLE
Multipart SSE support

### LINKED ISSUE #793: check and confirm SSE works with mutipart upload
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug
> /kind feature

**What happened**:

**What you expected to happen**:
To test multi part upload with encryption enabled on the bucket

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Gelato(release/branch) version:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by rhsakarpos**
**What this PR does / why we need it**:
This adds support for encryption during multipart upload

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #793 

**Special notes for your reviewer**:

**Release note**:

```
Uses stream encryption. Tested with 5MB plus object size, images, text files etc.
This PR has so many files, why ?
1. was running go fmt. by mistake, ran it on whole project, could not undo it

List of files with code changes

Makefile
initdb.sh
tidb.sql
tidb/bucket.go
 s3/pkg/service/bucket.go
 s3/pkg/service/multipart.go
 s3/pkg/service/object.go 
 s3/pkg/service/service.go
 s3/pkg/utils/encryptionutils.go



```

### CODE REVIEW

**Comment by rhsakarpos** [comment]
@kumarashit @sfzeng @nguptaopensds please review, thanks

File: s3/pkg/service/multipart.go [hunk 1]
```diff
@@ -260,7 +262,35 @@ func (s *s3Service) UploadPart(ctx context.Context, stream pb.S3_UploadPartStrea
 	md5Writer := md5.New()
 	data := &StreamReader{in: stream}
 	limitedDataReader := io.LimitReader(data, size)
-	dataReader := io.TeeReader(limitedDataReader, md5Writer)
+	var dataReader io.Reader
+	// encrypt if needed
+	if bucket.ServerSideEncryption.SseType == "SSE" {
+		/*byteArr, _ := ioutil.ReadAll(limitedDataReader)
```
Comment (kumarashit) [inline s3/pkg/service/multipart.go]: Please remove this code

File: s3/pkg/service/multipart.go [hunk 2]
```diff
@@ -402,17 +432,18 @@ func (s *s3Service) CompleteMultipartUpload(ctx context.Context, in *pb.Complete
...
+	mpures, mpuerr := sd.CompleteMultipartUpload(ctx, &pb.MultipartUpload{
 		Bucket:   bucketName,
 		Key:      objectKey,
 		UploadId: uploadId,
 		ObjectId: multipart.ObjectId,
 		Location: multipart.Metadata.Location,
 		Tier:     multipart.Metadata.Tier,
 	}, completeUpload)
-	if err != nil {
+	fmt.Println(mpures)
```
Comment (kumarashit) [inline s3/pkg/service/multipart.go]: Is this for debugging?

File: backend/cmd/main_test.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package main
```
Comment (kumarashit) [inline backend/cmd/main_test.go]: Is this file required?

File: s3/cmd/main_test.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package main
```
Comment (kumarashit) [inline s3/cmd/main_test.go]: Didn't get the purpose of this

File: s3/pkg/utils/main/mainencdec.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
+package main
```
Comment (kumarashit) [inline s3/pkg/utils/main/mainencdec.go]: Is this file a test utility? required?

File: s3/pkg/service/multipart.go (hunk 1 above)
Comment (rhsakarpos) [reply s3/pkg/service/multipart.go]: done

File: s3/pkg/service/multipart.go (hunk 2 above)
Comment (rhsakarpos) [reply s3/pkg/service/multipart.go]: yes, removed

File: backend/cmd/main_test.go (hunk 3 above)
Comment (rhsakarpos) [reply backend/cmd/main_test.go]: yes, this is the way to run the server, to get coverage for any kind of FT or automation testing

File: s3/cmd/main_test.go (hunk 4 above)
Comment (rhsakarpos) [reply s3/cmd/main_test.go]: yes, this is the way to run the server, to get coverage for any kind of FT or automation testing

File: s3/pkg/utils/main/mainencdec.go (hunk 5 above)
Comment (rhsakarpos) [reply s3/pkg/utils/main/mainencdec.go]: main file, for quick testing of encryption and decryption

**Comment by kumarashit** [comment]
Do we need the s3 binary?

**Comment by rhsakarpos** [comment]
> Do we need the s3 binary?

no, have removed it.

File: backend/cmd/main_test.go (hunk 3 above)
Comment (kumarashit) [reply backend/cmd/main_test.go]: Ok. I was confused as this code was not specific to SSE testing but generic

File: s3/cmd/main_test.go (hunk 4 above)
Comment (kumarashit) [reply s3/cmd/main_test.go]: Same as above

**Comment by kumarashit** [comment]
LGTM
